WcaOnRails/app/webpacker/lib/wca-live/attempts.js

```diff
@@ -1,23 +1,126 @@
+import _ from 'lodash';
+
 export const SKIPPED_VALUE = 0;
 export const DNF_VALUE = -1;
 export const DNS_VALUE = -2;
 
+function isComplete(attemptResult) {
+  return attemptResult > 0;
+}
+
+function isSkipped(attemptResult) {
+  return attemptResult === SKIPPED_VALUE;
+}
+
+function compareAttemptResults(attemptResult1, attemptResult2) {
+  if (!isComplete(attemptResult1) && !isComplete(attemptResult2)) return 0;
+  if (!isComplete(attemptResult1) && isComplete(attemptResult2)) return 1;
+  if (isComplete(attemptResult1) && !isComplete(attemptResult2)) return -1;
+  return attemptResult1 - attemptResult2;
+}
+
+function mean(x, y, z) {
+  return Math.round((x + y + z) / 3);
+}
+
+function meanOf3(attemptResults) {
+  if (!attemptResults.every(isComplete)) return DNF_VALUE;
+  return mean(...attemptResults);
+}
+
+function averageOf5(attemptResults) {
+  const [, x, y, z] = attemptResults.slice().sort(compareAttemptResults);
+  return meanOf3([x, y, z]);
+}
+
+/* See: https://www.worldcubeassociation.org/regulations/#9f2 */
+function roundOver10Mins(value) {
+  if (!isComplete(value)) return value;
+  if (value <= 10 * 6000) return value;
+  return Math.round(value / 100) * 100;
+}
+
 /**
- * Converts centiseconds to a human-friendly string.
+ * Returns the specified number of attempt results filling missing ones with 0.
  */
-function centisecondsToClockFormat(centiseconds) {
-  if (centiseconds == null) {
-    return '?:??:??';
+export function padSkipped(attemptResults, numberOfAttempts) {
+  return _.times(
+    numberOfAttempts,
+    (index) => (index < attemptResults.length ? attemptResults[index] : SKIPPED_VALUE),
+  );
+}
+
+/**
+ * Removes trailing skipped attempt results from the given list.
+ */
+export function trimTrailingSkipped(attemptResults) {
+  if (attemptResults.length === 0) return [];
+  if (attemptResults[attemptResults.length - 1] === SKIPPED_VALUE) {
+    return trimTrailingSkipped(attemptResults.slice(0, -1));
   }
-  if (!Number.isFinite(centiseconds)) {
-    throw new Error(
-      `Invalid centiseconds, expected positive number, got ${centiseconds}.`,
-    );
+  return attemptResults;
+}
+
+/**
+ * Returns the best attempt result from the given list.
+ *
+ * @example
+ * best([900, -1, 700]); // => 700
```

Comment:
I feel like this could benefit from some more examples of non-trivial cases like DNS, DNS, DNF and similar.